deducing-this.md

```diff
@@ -747,6 +836,106 @@ Keep in mind that there are a few more functions in P0798 that have this lead to
 
 For those that dislike returning auto in these cases, it is very easy to write a metafunction that matches the appropriate qualifiers from a type. Certainly simpler than copying and pasting code and hoping that the minor changes were made correctly in every case.
 
+## Recursive Lambdas
+
+This proposal also allows for an alternative solution to implementing a recursive lambda, since now we open up the possibility of allowing a lambda to reference itself:
+
+```
+// as proposed in [P0839]
+auto fib = [] self (int n) {
+    if (n < 2) return n;
+    return self(n-1) + self(n-2);
+};
+
+// this proposal
+auto fib = [](auto& this self, int n) {
+    if (n < 2) return n;
+    return self(n-1) + self(n-2);
+};
+```
+
+This simply works following the established rules. The call operator of the closure object can have an explicit object parameter too, so `self` in this example *is* the closure object.
+
+If the lambda would otherwise decay to a function pointer, `&self` shall have the value of that function pointer.
+
+### Expressions allowed for `self` in lambdas
+
+```
+  self(...);      // call with appropriate signature
+  decltype(self); // evaluates to the type of the lambda with the appropriate
+                  // cv-ref qualifiers
+  &self;          // the address of either the closure object or function pointer
+  std::move(self) // you're allowed to move yourself into an algorithm...
+  /* ... and all other things you're allowed to do with the lambda itself. */
+```
+
+Within lambda expressions, the `this` parameter still does not allow one to refer to the members of the closure object, which has no defined storage or layout, nor do its members have names. Instead it allows one to deduce the value category of the lambda and access its members -- including various call operators -- in the way appropriate for the value category.
+
+### Deducing derived objects for generic lambdas
+
+When we deduce the object parameter, it doesn't have to be the same type as the class - it could be a derived type. This isn't typically relevant for lambdas, which would just be used as is, but it does provide an interesting opportunity when it comes to visitation. One tool to make it easier to invoke `std::visit()` on a `std::variant` is to write an overload utility which ends up creating a new type that inherits from all of the provided types. Consider a scenario where we have a `Tree` type which is a `variant<Leaf, Node>`, where a `Node` contains two sub-`Tree`s. We could count the number of leaves in a `Tree` as follows:
+
+```
+int num_leaves(Tree const& tree) {
+    return std::visit(overload(
+        [](Leaf const&) { return 1; },
+        [](auto const& this self, Node const& n) {
+            return std::visit(self, n.left) + std::visit(self, n.right);
+        }
+    ), tree);
+}
+```
+
+In this example, `self` doesn't deduce as that inner lambda, it deduces at the overload object, so this works.
+
+## CRTP, without the C, R, or even T
```

Comment:
This made me emit a high-pitched noise that startled my wife. I demand forewarning for humor.